src/client/Microsoft.Identity.Client/Internal/Requests/RequestBase.cs

```diff
@@ -210,7 +210,8 @@ protected async Task<AuthenticationResult> CacheTokenResponseAndCreateAuthentica
                 atItem, 
                 idtItem, 
                 AuthenticationRequestParameters.AuthenticationScheme,
-                AuthenticationRequestParameters.RequestContext.CorrelationId);
+                AuthenticationRequestParameters.RequestContext.CorrelationId,
+                msalTokenResponse.TokenSource == TokenSource.None? TokenSource.IdentityProvider : msalTokenResponse.TokenSource);
```

Comment:
Why can't msalTokenResponse.TokenSource always have the correct value?

Reply:
There are methods in the MsalTokenResponse class to build response for broker. So I am setting the value of token source to broker there. https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1933/files#diff-ea5df3781a2dc2119f53390344fbfbd7R136

This method will either have the token source as Broker or Identity Provider. So instead I can check if Broker is set and otherwise set Identity Provider.